pkg/controller/machine/machine_controller_suite_test.go

```diff
@@ -18,53 +18,94 @@ package machine
 
 import (
 	"context"
+	"flag"
 	"log"
-	"os"
 	"path/filepath"
 	"testing"
+	"time"
+
+	. "github.com/onsi/ginkgo/v2"
+	. "github.com/onsi/gomega"
 
 	machinev1 "github.com/openshift/api/machine/v1beta1"
 	"k8s.io/client-go/kubernetes/scheme"
 	"k8s.io/client-go/rest"
+	"k8s.io/klog/v2/textlogger"
+	"sigs.k8s.io/controller-runtime/pkg/client"
 	"sigs.k8s.io/controller-runtime/pkg/envtest"
-	"sigs.k8s.io/controller-runtime/pkg/manager"
+	logf "sigs.k8s.io/controller-runtime/pkg/log"
+)
+
+func init() {
+	textLoggerConfig := textlogger.NewConfig()
+	textLoggerConfig.AddFlags(flag.CommandLine)
+	logf.SetLogger(textlogger.NewLogger(textLoggerConfig))
+
+	// Register required object kinds with global scheme.
+	_ = machinev1.Install(scheme.Scheme)
+}
+
+const (
+	timeout = time.Second * 10
 )
 
 var (
-	cfg *rest.Config
-	ctx = context.Background()
+	cfg       *rest.Config
+	ctx       = context.Background()
+	k8sClient client.Client
+	testEnv   *envtest.Environment
 )
 
-func TestMain(m *testing.M) {
-	t := &envtest.Environment{
-		CRDDirectoryPaths: []string{filepath.Join("..", "..", "..", "install")},
+func TestMachineController(t *testing.T) {
+	RegisterFailHandler(Fail)
+
+	RunSpecs(t, "Machine Controller Suite")
+}
+
+var _ = BeforeSuite(func(ctx SpecContext) {
+	By("bootstrapping test environment")
+	testEnv = &envtest.Environment{
+
+		CRDInstallOptions: envtest.CRDInstallOptions{
+			Paths: []string{
+				filepath.Join("..", "..", "..", "vendor", "github.com", "openshift", "api", "machine", "v1beta1", "zz_generated.crd-manifests", "0000_10_machine-api_01_machines-CustomNoUpgrade.crd.yaml"),
+			},
+		},
+	}
+
+	var err error
+	cfg, err = testEnv.Start()
+	Expect(err).ToNot(HaveOccurred())
+	Expect(cfg).ToNot(BeNil())
+
+})
+
+var _ = AfterSuite(func() {
+	By("tearing down the test environment")
+	Expect(testEnv.Stop()).To(Succeed())
+})
+
+func StartEnvTest(t *testing.T) func(t *testing.T) {
```

Comment:
Since this is kind of duplicating some of the logic above, is there a way to share them?

Perhaps have the func return the testEnv? And then you can `defer func() { testenv.Stop() }` in the non ginkgo formatted tests?

Reply:
Bump, don't think this one has been looked at yet